File: tree/102/102.py
#solution 1
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        levels=[]
        if root is None: return levels
        def helper(node,level):
            if len(levels)==level:levels.append([])
            levels[level].append(node.val)
            if node.left: helper(node.left,level+1)
            if node.right: helper(node.right,level+1)
        helper(root,0)
        return levels

class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        levels=[]
        if root is None: return levels
        queue=collections.deque([root])
        level=0
        while queue:
            levels.append([])
            level_len=len(queue)
            for i in range(level_len):
                node=queue.popleft()
                levels[level].append(node.val)
                if node.left: queue.append(node.left)
                if node.right: queue.append(node.right)
            level+=1
        return levels
                 

File: tree/102/102.js
//solution 1
var levelOrder = function (root) {
  let levels = [];
  if (!root) return levels;
  const helper = (node, level) => {
    if (levels.length === level) {
      levels.push([]);
    }
    levels[level].push(node.val);
    if (node.left) helper(node.left, level + 1);
    if (node.right) helper(node.right, level + 1);
  };
  helper(root, 0);
  return levels;
};

//solution 2
var levelOrder = (root) => {
  let levels = [];
  let queue = [];
  if (root) queue.push(root);
  let level = 0;
  while (queue.length !== 0) {
    levels.push([]);
    const levelLen = queue.length;
    //can't directly use queue.length here cuz queue's length changes everytime when you pop the head node
    for (let i = 0; i < levelLen; i++) {
      const node = queue.shift();
      levels[level].push(node.val);
      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
    }
    level++;
  }
  return levels;
};
